[#1178] feat(trino-connector) Support create/drop catalog in Trino.

diqiu50 opened the pull request:
### What changes were proposed in this pull request?

Support create/drop catalog in Trino.

```text
-- Call stored procedures with position.
call gravitino.system.create_catalog(
    'mysql',
    'jdbc-mysql',
    Map(
        Array('jdbc-url', 'jdbc-user', 'jdbc-password', 'jdbc-driver'),
        Array('jdbc:mysql:192.168.164.4:3306?useSSL=false', 'trino', 'ds123', 'com.mysql.cj.jdbc.Driver')
    )
)
call gravitino.system.drop_datalog('mysql');

-- Call stored procedures with name.
call gravitino.system.create_catalog(
    catalog =>'mysql',
    provider => 'jdbc-mysql',
    properties => Map(
        Array('jdbc-url', 'jdbc-user', 'jdbc-password', 'jdbc-driver'),
        Array('jdbc:mysql:192.168.164.4:3306?useSSL=false', 'trino', 'ds123', 'com.mysql.cj.jdbc.Driver')
    ),
    ignore_exist => true
)

call gravitino.system.drop_datalog(
    catalog => 'mysql'
    ignore_not_exist => true
);
```

### Why are the changes needed?

Feature: #1178

### How was this patch tested?

UT and IT

yuqi1129 commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -44,6 +47,12 @@ public String getMetalake() {
+  boolean isStatic() {
+    // The 'isStatic' is indicated whether the connector
+    // is user-configured within Trino or loaded from the Graviton server
+    return config.getOrDefault("gravitino.internal", "false").equals("false");
```
Please also change the configuration document for the Trino connector.

diqiu50 replied:
No, you don't need to set gravitino.internal. That's not a configuration.
It's only generated in `CatalogInjector.createCatalogProperties`.

jerryshao replied:
What is the usage of this `isStatic()`? You should have a better name or have a better explanation.

diqiu50 commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java:
```diff
@@ -66,17 +70,26 @@ public Connector create(
+      if (Strings.isNullOrEmpty(metalake)) {
+        throw new TrinoException(GRAVITINO_METALAKE_NOT_EXISTS, "No gravitino metalake selected");
+      }
+      catalogConnectorManager.addMetalake(metalake);
```
We can create multiple configuration files in the Trino catalog directory. Subsequently, we can have two Gravitino connectors to use two metalakes.

For example:

```
root@trino-01:/usr/local/trino-server-426/etc/catalog# ls
gravitino.properties  gravitino1.properties  jmx.properties  mysql.properties
```

gravitino.properties :
```
connector.name=gravitino
gravitino.url=http://localhost:8090
gravitino.metalake=test
```
gravitino1.properties 
```
connector.name=gravitino
gravitino.url=http://localhost:8090
gravitino.metalake=test1
```

diqiu50 commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:
```diff
@@ -103,20 +109,21 @@ public void start() {
-        LOG.error("Load Metalake {} failed.", usedMetalake, e);
-        return;
-      }
+      for (String usedMetalake : usedMetalakes) {
```
The typical metadata lake usually has only one or two. For now, we don't need to consider it.

yuqi1129 commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:
```diff
@@ -103,20 +109,21 @@ public void start() {
-      LOG.debug("Load metalake: " + usedMetalake);
-      loadCatalogs(metalake);
-      // TODO (yuhui) need to handle metalake dropped.
+        LOG.debug("Load metalake: " + usedMetalake);
```
It's not advisable to use `debug` level here.

yuqi1129 commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -44,6 +47,12 @@ public String getMetalake() {
 
+  boolean isStatic() {
+    // The 'isStatic' is indicated whether the connector
+    // is user-configured within Trino or loaded from the Graviton server
```
Graviton -> Gravitino

diqiu50 commented:
@jerryshao  Please take a review.

yuqi1129 commented:
> hao Please take a review.

I 've already reviewed it. It's fine with me. Please take some time to view it. Thanks.

diqiu50 commented:
> With this, does user still need to configure metalake in trino connector?

Yes，the Trino connector will only load the configured metadata lakes."

jerryshao commented:
So it the configured metalake is "test", and with call procedure you created a new one "test1", can "test1" be used by this connector, or user can still use the metalake "test"?

jerryshao commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:
```diff
@@ -20,8 +30,85 @@
+              new Procedure.Argument("METALAKE", VARCHAR),
+              new Procedure.Argument("CATALOG", VARCHAR),
+              new Procedure.Argument("PROVIDER", VARCHAR),
+              new Procedure.Argument("PROPERTIES", VARCHAR, false, "{}"),
```
Can it be Map type?

diqiu50 replied:
I think the map type in Trino a bit challenging to use, the usage is as follows:
```
call gravitino.system.createCatalog('test', 'jdbc-mysql7', 'jdbc-mysql', MAP(ARRAY['jdbc-url', 'jdbc-user', 'jdbc-password','jdbc-dri
ver'], ARRAY['jdbc:mysql://192.168.164.4:3306?useSSL=false', 'mysql', 'mysql', 'com.mysql.cj.jdbc.Driver']));
```
Using a JSON string might align better with our preferences.
```
call gravitino.system.createCatalog('test', 'jdbc-mysql', 'jdbc-mysql', '{"jdbc-url":"jdbc:mysql://192.168.164.4:3306?useSSL=false","jdbc-user":"mysql","jdbc-password":"mysql", "jdbc-driver":"com.mysql.cj.jdbc.Driver"}');
```

jerryshao replied:
I found that in https://iceberg.apache.org/docs/latest/spark-procedures/#examples-8, there're some call procedures that have required properties as map type. So I'm inclined to use map type. Though it is not so straight to write, but the definition of map type is much more clear than a vague varchar.

diqiu50 commented:
> So if the configured metalake is "test", and with call procedure you created a new one "test1", can "test1" be used by this connector, or user can still use the metalake "test"?

If 'test' metadata lake is configured and 'test1' metadata lake is not configured, users can still utilize the 'test' metadata lake. Trino will not load the catalog for the 'test1' metadata lake.

The primary purpose of creating a metalake is to allow users to create configured metalakes through Trino. Users can configure multiple Gravitino connectors, each using a different metadata lake.

diqiu50 commented:
@jerryshao 
I have already removed the stored procedures for creating/dropping the `metalake` and added relevant documentation.

Regarding the attributes section in 'create catalog,' Trino doesn't have a better type of solution to this problem. Using a JSON format or a format similar to 'key1=value1;key2=value2;...' could be relatively more user-friendly.

Later, if we make changes to Trino, we can define specific syntax for the 'create catalog' statement to address the issues caused by using plain strings.

```
CREATE  CATALOG NAME by PROVIDER WITH (
	key1:value1,
	key2:value2,
);
```
Currently, I use JSON, which makes implementation easier.

jerryshao commented:
@diqiu50 would you please update the PR title and description to reflect the current status of PR?

jerryshao commented on docs/trino-connector/supported-catalog.md:
```diff
@@ -15,7 +15,43 @@ The catalogs currently supported by the Gravitino connector are as follows:
+
+- CATALOG: The catalog name to be created.
+- PROVIDER: The catalog provider, currently only supports `hive`, `lakehouse-iceberg`, `jdbc-mysql`, `jdbc-postgresql`.
+- PROPERTIES: The properties of the catalog, the format is a json string. like `{"key1":"value1","key2":"value2"}`.
```
Should you also specify which catalog properties are supported or provided by referring to the specific docs?

jerryshao replied:
Also can you follow this Iceberg doc to make it as a table for properties. https://iceberg.apache.org/docs/latest/spark-procedures/

jerryshao commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:
```diff
@@ -20,8 +30,52 @@
+          List.of(
+              new Procedure.Argument("CATALOG", VARCHAR),
+              new Procedure.Argument("IGNORE NOT EXIST", BOOLEAN, false, true));
+      procedure = new Procedure("system", "dropCatalog", arguments, dropCatalog);
```
Is trino case insensitive, what will be happened if user write as "DROPCATALOG"? I would suggest to use "drop_catalog", "create_catalog".

diqiu50 replied:
Ok

jerryshao commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:
```diff
@@ -56,7 +67,7 @@ public class CatalogConnectorManager {
   private GravitinoClient gravitinoClient;
   private GravitinoConfig config;
-  private String usedMetalake;
+  private final Set<String> usedMetalakes = new HashSet<>();
```
Is this change still required?

diqiu50 replied:
We need to support configuring multiple connectors, each using a different metastore.

jerryshao commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:
```diff
@@ -191,4 +206,110 @@ public void shutdown() {
+    }
+  }
+
+  public void createMetalake(String name, boolean ignoreExist) {
```
I think this is not required, right?

diqiu50 replied:
Yes

diqiu50 commented:
@jerryshao 
I have changed the properties to a map type and updated the documents.
Please take a review.

jerryshao commented on docs/trino-connector/supported-catalog.md:
````diff
@@ -15,7 +15,74 @@ The catalogs currently supported by the Gravitino connector are as follows:
+
+Create catalog:
+
+```text
````
Can you please change this to `sql`?

jerryshao commented on trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:
```diff
@@ -20,8 +33,53 @@
+      Procedure procedure = new Procedure("system", "create_catalog", arguments, createCatalog);
+      procedures.add(procedure);
+
+      // call gravitino.system.dropCatalog(metalake, catalog, ignoreNotExist)
```
I think you need to update the comment here.

jerryshao commented on docs/trino-connector/supported-catalog.md:
````diff
@@ -52,7 +52,7 @@ So you need to use the following SQL to call them in the `trino-cli`:
 You can run the following SQL to create a catalog named `mysql` with `jdbc-mysql` provider.
 
-```text
+```sql
````
It's not just here, also above, please double check the whole doc.
